libr/include/r_vector.h

```diff
@@ -7,93 +7,183 @@ extern "C" {
 #endif
 
 /*
-  Insert functions returns NULL if reallocation happens and fails,
-  returns the address of newly inserted element if succeeded.
-
-  Delete functions returns the deleted element.
-  Callers should destruct it if necessary.
-
-  Usage:
-  int i;
-  void **it;
-
-  // heap allocated RVector
-  RVector *v = r_vector_new ();
-  if (!v) goto err_new;
-  if (!r_vector_push (v, a)) goto err_push;
-  if (!r_vector_insert (v, 0, b)) goto err_insert;
-  for (i = 0; i < v->len; i++) {
-    (void)v->a[i]; // Public members a and len are encouraged to access
-  }
-  r_vector_foreach (v, it) {
-    (void)*it;
-  }
-  // pass function pointer `elem_free`
-  r_vector_free (v, NULL);
-
-  // stack allocated RVector
-  RVector v = {0};
-  // r_vector_init (&v); // v.a = NULL; v.len = v.capacity = 0;
-  r_vector_push (&v, (void *)1);
-  assert (v.len == 1 && v.capacity == 10);
-  // for stack allocated RVector, use r_vector_clear instead of r_vector_free
-  r_vector_clear (&v, NULL);
- */
+  RVector can contain arbitrarily sized elements.
+  RPVector uses RVector internally and always contains void *s.
+
+  Hint: remove/pop functions do not reduce the capacity.
+        Call r_(p)vector_shrink explicitly if desired.
+*/
+
+typedef int (*RPVectorComparator)(const void *a, const void *b);
+typedef void (*RVectorFree)(void *e, void *user);
+typedef void (*RPVectorFree)(void *e);
 
 typedef struct r_vector_t {
-	void **a;
-	int len;
-	int capacity;
+	void *a;
+	size_t len;
+	size_t capacity;
+	size_t elem_size;
 } RVector;
 
-typedef int (*RVectorComparator)(const void *a, const void *b);
-typedef void (*RVectorFree)(void *e);
+typedef struct r_pvector_t {
+	RVector v;
+	RPVectorFree free;
+} RPVector;
```

Comment:
Yeah if we embed free (and free_user) inside RVector, we could get rid of RPVector.

But I think it still makes sense to have a separate type `RPVector` that is not just a typedef, so you don't accidently call r_vector_* functions on an RPVector and vice versa, which could be dangerous, or at least the compiler will warn you.
A simple `typedef struct r_pvector_t { RVector v; } RPVector;` would do the trick. What do you think?

Reply:
why do we need a separate type at all? a RPVector is actually just a RVector (if we embed the `free` field)

Reply:
Exactly, we don't. But it would be nice for type safety since. For instance `r_vector_push` and `r_pvector_push` take the same parameters, but the vector one reads *x and the pvector one uses x directly.